network/tls.go

```diff
@@ -0,0 +1,171 @@
+package network
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"time"
+
+	"fmt"
+
+	"github.com/dedis/cothority/log"
+)
+
+/*
+Implementation of a TLS-host to allow for tls-connections. Only the server
+will authenticate to the client.
+*/
+
+// NewTLSRouter returns a new Router using TLSHost as the underlying Host.
+func NewTLSRouter(sid *ServerIdentity, key TLSKeyPEM) (*Router, error) {
+	h, err := NewTLSHost(sid, key)
+	if err != nil {
+		return nil, err
+	}
+	r := NewRouter(sid, h)
+	return r, nil
+}
+
+// TLSHost implements the Host interface using TLS connections.
+type TLSHost struct {
+	*TLSListener
+	addr Address
+	cert TLSCertPEM
+	key  TLSKeyPEM
+}
+
+// NewTLSHost returns a new Host using TLS connection based type.
+func NewTLSHost(si *ServerIdentity, key TLSKeyPEM) (*TLSHost, error) {
+	h := &TLSHost{
+		addr: si.Address,
+		cert: si.Cert,
+		key:  key,
+	}
+	var err error
+	h.TLSListener, err = NewTLSListener(si, key)
+	return h, err
+}
+
+// Connect establishes a connection to the remote host given in addr.
+func (h *TLSHost) Connect(si *ServerIdentity) (Conn, error) {
+	log.Lvl3(h.addr, "connecting to", si.Address)
+	addr := si.Address
+	switch addr.ConnType() {
+	case TLS:
+		c, err := NewTLSConn(si)
+		return c, err
+	}
+	return nil, fmt.Errorf("Don't support connection: %s", addr.ConnType())
+}
+
+// TLSListener implements the Host-interface using TLS as a communication channel.
+type TLSListener struct {
+	*TCPListener
+}
+
+// NewTLSListener returns a TLSListener. This function binds to the given
+// address.
+// It returns the listener and an error if one occurred during
+// the binding.
+// A subsequent call to Address() gives the actual listening
+// address which is different if you gave it a ":0"-address.
+func NewTLSListener(si *ServerIdentity, key TLSKeyPEM) (*TLSListener, error) {
+	log.Lvl3("Starting to listen on", si)
+	addr := si.Address
+	if addr.ConnType() != TLS {
+		return nil, errors.New("TLSListener can't listen on non-TLS address")
+	}
+	t := &TLSListener{
+		TCPListener: &TCPListener{
+			quit:         make(chan bool),
+			quitListener: make(chan bool),
+		},
+	}
+	global, _ := GlobalBind(addr.NetworkAddress())
```

Comment:
I was planning to re-use the TCPConn to upgrade to TLSConn so to avoid duplication of code. You can do it like explained [here](https://stackoverflow.com/questions/13110713/upgrade-a-connection-to-tls-in-go)